### PR TITLE
Rewrite without dolittle

### PR DESCRIPTION
**Pull request by srudaa**
Rewrite and rename the NMEA connector based on new framework (https://github.com/RaaLabs/Edge).

### CODE REVIEW

File: Specifications/Drivers/EventParsedInstanceFactory.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+using TechTalk.SpecFlow.Assist;
+
+namespace RaaLabs.Edge.Connectors.NMEA.Specs.Drivers
+{
+    class EventParsedInstanceFactory : IEventInstanceFactory<EventParsed>
+    {
+        public EventParsed FromTableRow(TableRow row)
+        {
+
+            var eventParsed = new events.EventParsed
```
Comment (thotokraa) [inline Specifications/Drivers/EventParsedInstanceFactory.cs:18]: This expression can possibly be replaced with the following code:
```csharp
var eventParsed = row.CreateInstance<events.EventParsed>();
```

File: Specifications/Drivers/IEventInstanceFactory.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using TechTalk.SpecFlow;
+
+namespace RaaLabs.Edge.Connectors.NMEA.Specs.Drivers
+{
+    interface IEventInstanceFactory<T>
```
Comment (thotokraa) [inline Specifications/Drivers/IEventInstanceFactory.cs:10]: This interface is part of `RaaLabs.Edge.Testing` from version 1.1.0.

File: Specifications/Drivers/IProducedEventVerifier.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using TechTalk.SpecFlow;
+
+namespace RaaLabs.Edge.Connectors.NMEA.Specs.Drivers
+{
+    interface IProducedEventVerifier<T>
```
Comment (thotokraa) [inline Specifications/Drivers/IProducedEventVerifier.cs:10]: This interface is part of `RaaLabs.Edge.Testing` from version 1.1.0.

File: Specifications/Drivers/TypeMapper.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using RaaLabs.Edge.Connectors.NMEA;
+
+namespace RaaLabs.Edge.Connectors.NMEA.Specs.Drivers
+{
+    public class TypeMapping : Dictionary<string, Type>
```
Comment (thotokraa) [inline Specifications/Drivers/TypeMapper.cs:10]: This class is part of `RaaLabs.Edge.Testing` from version 1.1.0. It should be replaced with a hook for injecting the type mappings into that class, as instructed in the readme file for the testing library.

File: Specifications/Features/SentenceParser.feature [hunk 5]
```diff
@@ -0,0 +1,15 @@
+Feature: SentenceParser
+
+    Scenario: Parse unsupported sentences
+        When sentence "$AAAAA,5.2,0.0,10.0*63" is parsed
+        Then the exception "UnsupportedSentenceException" is thrown with message "Talker 'AA' with identifier 'AAA' is not supported in sentence '$AAAAA,5.2,0.0,10.0*63'"
```
Comment (thotokraa) [inline Specifications/Features/SentenceParser.feature:5]: In my opinion, testing the string content of an exception is not the best way to test the outcome from a test, although it is very convenient. We could discuss this a bit further before merging.

File: Specifications/Features/StateHandler.feature [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        Then the following events of type NMEADatapointOutput is produced
+            | Source | Tag        | Value | Timestamp     |
+            | NMEA   | WaterDepth | 5.2   | 1617915312000 |
+
+    Scenario: Handling incoming parsed events with different priority. Where first event received has highest priority.
+        When the following events of type EventParsed is produced
+            | Identifier | Tag         | Value | Timestamp     |
+            | HEHDT      | HeadingTrue | 5.0   | 1617915312000 |
+            | HEVHW      | HeadingTrue | 5.2   | 1617915312000 |
+        Then the following events of type NMEADatapointOutput is produced
```
Comment (thotokraa) [inline Specifications/Features/StateHandler.feature:22]: As the testing framework works right now, we don't test that the number of events produced is exactly equal to the number of rows in this table, which means that this test would probably pass even if you passed HEVHW before HEHDT. This needs to be fixed in the Testing library.

File: Specifications/Features/StateHandler.feature [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    Scenario: Handling incoming parsed events with different priority. Where second event received has highest priority.
+        When the following events of type EventParsed is produced
+            | Identifier | Tag         | Value | Timestamp     |
+            | HEVHW      | HeadingTrue | 5.2   | 1617915312000 |
+            | HEHDT      | HeadingTrue | 5.0   | 1617915312000 |
+        Then the following events of type NMEADatapointOutput is produced
+            | Source | Tag         | Value | Timestamp     |
+            | NMEA   | HeadingTrue | 5.2   | 1617915312000 |
+            | NMEA   | HeadingTrue | 5.0   | 1617915312000 |
```
Comment (thotokraa) [inline Specifications/Features/StateHandler.feature:34]: Could you add some tests for what should happen to incoming data points when the current data points are stale (produced more than 60 seconds ago)?

File: Specifications/Hooks/InstanceFactoryProvider.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using TechTalk.SpecFlow;
+
+namespace RaaLabs.Edge.Connectors.NMEA.Specs.Hooks
+{
+    [Binding]
+    class InstanceFactoryProvider
```
Comment (thotokraa) [inline Specifications/Hooks/InstanceFactoryProvider.cs:14]: This class is not needed if you use `RaaLabs.Edge.Testing` v1.1.0.

File: Specifications/Hooks/LoggerProvider.cs [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+using System.Linq;
+using System.Text;
+using TechTalk.SpecFlow;
+using Serilog;
+using BoDi;
+
+namespace RaaLabs.Edge.Connectors.NMEA.Specs.Hooks
+{
+    [Binding]
+    public sealed class LoggerProvider
```
Comment (thotokraa) [inline Specifications/Hooks/LoggerProvider.cs:12]: This class is not needed if you use `RaaLabs.Edge.Testing`.

File: Specifications/Hooks/SentenceParserProvider.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+            _container = container;
+        }
+
+        [BeforeScenario]
+        public void SetupSentenceParserProvider()
+        {
+            var assembly = typeof(ISentenceFormat).Assembly;
+            var sentenceFormats = assembly.GetTypes().Where(type => typeof(ISentenceFormat).IsAssignableFrom(type) && !type.IsInterface);
+            _container.RegisterFactoryAs<IEnumerable<ISentenceFormat>>(_ => sentenceFormats.Select(sentenceFormat => (ISentenceFormat) Activator.CreateInstance(sentenceFormat)) );
+            //var registerFunction = GetType().GetMethod("RegisterSentenceFormat");
```
Comment (thotokraa) [inline Specifications/Hooks/SentenceParserProvider.cs:29]: Can we remove these lines now?

File: Specifications/Steps/HandlerSteps.cs [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using TechTalk.SpecFlow;
+using FluentAssertions;
+
+namespace RaaLabs.Edge.Connectors.NMEA.Specs.Steps
+{
+    [Binding]
+    public sealed class HandlerSteps
```
Comment (thotokraa) [inline Specifications/Steps/HandlerSteps.cs:14]: This class is now provided in the `RaaLabs.Edge.Testing` library.

File: Specifications/specflow.json [hunk 12]
```diff
@@ -0,0 +1,3 @@
+{
+    "allowDebugGeneratedFiles": true
```
Comment (thotokraa) [inline Specifications/specflow.json:2]: To use `RaaLabs.Edge.Testing`, you also need to add `stepAssemblies` field to this file, as specified in the testing framework readme.

**Review by thotokraa — CHANGES_REQUESTED** [review]

File: Source/Connector.cs [hunk 13]
```diff
@@ -1,148 +1,132 @@
...
-            switch (_configuration.Protocol)
+            while (true)
             {
-                case Protocol.Tcp: ConnectTcp(); break;
-                case Protocol.Udp: ConnectUdp(); break;
-                default: _logger.Error("Protocol not defined"); break;
+                _logger.Information("Setting up TCP connector");
+                var policy = Policy
+                    .Handle<Exception>()
+                    .WaitAndRetryForeverAsync(retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)),
```
[the remark this answers is not in the document]
Comment (srudaa) [reply Source/Connector.cs:54]: Should we?

File: Source/Connector.cs [hunk 14]
```diff
@@ -1,148 +1,132 @@
...
-                                            _logger.Error(ex, "Unable to parse sentence");
-                                        }
-                                    }
-                                    break;
-                            }
-                            if (started && !skip) sentenceBuilder.Append(character);
-                            skip = false;
-                        }
+                        await DoWithTimeout(reader.MoveNextAsync(), 3_000);
+                        var sentence = reader.Current;
```
[the remark this answers is not in the document]
Comment (srudaa) [reply Source/Connector.cs:87]: I did not get this to work. However, if it times out the socket will close and a new reader will be defined?

File: Specifications/Features/StateHandler.feature (hunk 6 above)
Comment (srudaa) [reply Specifications/Features/StateHandler.feature:22]: Fixed in testing library v1.1.1

File: Specifications/Drivers/EventParsedInstanceFactory.cs (hunk 1 above)
Comment (srudaa) [reply Specifications/Drivers/EventParsedInstanceFactory.cs:18]: Did not work. Got a lot of errors regarding floats. Will keep it as it is.

**Comment by srudaa** [comment]
@thotokraa could you have another look. I have pushed fixes based on your awesome feedback.

File: Source/NMEAStreamReader.cs [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+using System.Text;
+using System.IO;
+using System.Linq;
+
+namespace RaaLabs.Edge.Connectors.NMEA
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public class NMEAStreamReader
```
Comment (thotokraa) [inline Source/NMEAStreamReader.cs:15]: Add `static` here.

**Review by thotokraa — CHANGES_REQUESTED** [review]
Please fix the last code smell before merging.

**Review by thotokraa — APPROVED** [review]

**Pull request merged**